pkg/policy/selectorcache.go

```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
 	l.setSelections(&selections)
 }
 
-type fqdnSelector struct {
-	labelIdentitySelector
-	// fqdnSelections map[string]identity.NumericIdentity // identity.String(): "cidr:1.1.1.1" -> identity of 1.1.1.1
+func (l *fqdnSelector) setSelections(selections *[]identity.NumericIdentity) {
+	if len(*selections) > 0 {
+		atomic.StorePointer(&l.selections, unsafe.Pointer(selections))
+	} else {
+		atomic.StorePointer(&l.selections, unsafe.Pointer(&emptySelection))
+	}
+}
+
+// updateSelections updates the immutable slice representation of the
+// cached selections after the cached selections have been changed.
+//
+// lock must be held
+func (l *fqdnSelector) updateSelections() {
+	selections := make([]identity.NumericIdentity, len(l.cachedSelections))
+	i := 0
+	for nid := range l.cachedSelections {
+		selections[i] = nid
+		i++
+	}
+	// Sort the numeric identities so that the map iteration order
+	// does not matter. This makes testing easier, but may help
+	// identifying changes easier also otherwise.
+	sort.Slice(selections, func(i, j int) bool {
+		return selections[i] < selections[j]
+	})
+	l.setSelections(&selections)
+}
+
+// UpdateFQDNSelector updates the mapping of fqdnKey (the FQDNSelector from a
+// policy rule as a string) to to the provided list of identities. If the contents
+// of the cachedSelections differ from those in the identities slice, all
+// users are notified.
+func (sc *SelectorCache) UpdateFQDNSelector(fqdnKey string, identities []identity.NumericIdentity) {
```

Comment:
does the fdqnKey really needs to be a string? What are the implications of the number of memory allocations for this string for each policy recalculation? Why can't this be a `FQDNSelector` instead? Even in tests you are using `FQDNSelector` [1]

Reply:
Good idea, converting to `api.FQDNSelector`.

Reply:
@ianvernon What I meant here was using `FQDNSelector` directly the key of the map, not to pass the structure and they use a `.String()`

Reply:
The `SelectorCache` also uses the string representation of `api.EndpointSelector` as the key in the map, so we can't use `api.FQDNSelector` as the key directly. Do you want the key to be an interface type?

What about caching the contents of `String()` within a selector once we use it in the `SelectorCache`? This would make calls to `String()` only have an upfront cost of an initial computation, and since selectors are immutable, their string representation would always be the same. 

@jrajahalme do you have anything you'd like to add here?

Reply:
`api.EndpointSelector` as a map key will be different for each `policy import` event if the selectors are the same. This is due to the `api.EndpointSelector` only containing pointers, which will be different for each re-creation of the same selector from JSON, etc. I haven't looked at the details of the `api.FQDNSelector`, but it could have the same issue.

Caching the string representation in the `api.Selector` looks like a sound proposal to me.